File: tests/unittests/test_triangularmesh.py
import numpy as np

def create_reference_data():
    from mtlheartmesh.trisurf import TriSurf
    basename = '../data/test_mesh3d'
    S = TriSurf.read_off(basename+'.off')
    with open(basename+'.nbound_ref', 'wt') as file:
        file.write(f'{len(S.boundary_vertices)}')
    with open(basename+'.radius_ref', 'wt') as file:
        file.write(f'{S.face_altitudes.min():.9f}')
    C = S.face_vertex_connectivity
    with open(basename+'.inctri_ref', 'wt') as file:
        for i in range(C.shape[0]):
            file.write(str(C[i].nonzero()[1])[1:-1]+'\n')
    S.adjacent_faces.tofile(basename+'.adjtri_ref', sep=' ')
    B = np.unique(np.concatenate(S.boundary_vertices))
    B.tofile(basename+'.bound_ref', sep=' ')
    with open(basename+'.neightri_ref', 'wt') as file:
        for i in range(S.nt):
            L = np.unique(np.concatenate([C[k].nonzero()[1]
                                          for k in S.tri[i]]))
            file.write(str(L)[1:-1]+'\n')

def test_geometric_calculations():
    basename = 'test_mesh3d'
    basename_ref = '../data/test_mesh3d'

    # check radius
    h = np.fromfile(basename+'.radius', sep=' ')
    h_ref = np.fromfile(basename_ref+'.radius_ref', sep=' ')
    assert(abs(h - h_ref) < 1e-5)
    
    # check v_incident_tri
    with open(basename+'.inctri') as file:
        with open(basename_ref+'.inctri_ref') as file_ref:
            for i, (line, line_ref) in enumerate(zip(file.readlines(),
                                                     file_ref.readlines())):
                items = sorted([int(k) for k in line.split(' ')[:-1]])
                items_ref = sorted([int(k) 
                                    for k in line_ref.strip().split(' ') 
                                    if k])
                assert np.all(np.array(items) == items_ref)

    # check t_adjacent_tri
    A = np.fromfile(basename+'.adjtri', sep=' ').reshape((-1, 3))
    A_ref = np.fromfile(basename_ref+'.adjtri_ref', sep=' ').reshape((-1, 3))
    assert np.all(A.sort(1) == A_ref.sort(1))

    #check is_boundary_vertex
    B = np.fromfile(basename+'.bound', sep=' ').astype(int)
    B_ref = np.fromfile(basename_ref+'.bound_ref', sep=' ').astype(int)
    assert np.all(np.array(B) == B_ref)

    # check t_neigh_tri
    with open(basename+'.neightri', 'rt') as file:
        with open(basename_ref+'.neightri_ref') as file_ref:
            for i, (line, line_ref) in enumerate(zip(file.readlines(),
                                                     file_ref.readlines())):
                items = sorted([i] + [int(k) for k in line.split(' ')[:-1]])
                items_ref = sorted([int(k) 
                                    for k in line_ref.strip().split(' ') 
                                    if k])
                assert np.all(np.array(items) == items_ref)


#create_reference_data()
test_geometric_calculations()

File: tests/unittests/test_triangularmesh.cpp
#include "engine.cpp"
#include "readmesh.cpp"

int main()
{
    char fname[] = "../data/test_mesh3d.off";
    int nv, nt;
    double *ver, *orient;
    int *tri;
    read_off(fname, nv, nt, ver, tri);
    dummy_orient(nt, orient);

    TriangularMesh S;
    S.initialize(nv, nt, ver, tri, orient, 0, 0);

    FILE* file = fopen("test_mesh3d.nbound", "w");
    fprintf(file, "%.9f", 2 - S.euler_characteristic);
    fclose(file);

    file = fopen("test_mesh3d.radius", "w");
    fprintf(file, "%.9f", S.triangle_min_altitude);
    fclose(file);

    // check v_incident_tri
    file = fopen("test_mesh3d.inctri", "w");
    for (int i=0;i<S.nv;i++) {
        for (int j=0;j<S.v_nb_incident_tri[i];j++)
            fprintf(file, "%d ", S.v_incident_tri[S.v_maxtri*i+j]);
        fprintf(file, "\n");
    }
    fclose(file);

    // check t_adjacent_tri
    file = fopen("test_mesh3d.adjtri", "w");
    for (int i=0;i<S.nt;i++) {
        fprintf(file, "%d %d %d\n", S.t_adjacent_tri[3*i], 
                                    S.t_adjacent_tri[3*i+1], 
                                    S.t_adjacent_tri[3*i+2]);
    }
    fclose(file);

    // check is_boundary_vertex
    file = fopen("test_mesh3d.bound", "w");
    for (int i=0;i<S.nv;i++) if (S.is_boundary_vertex[i]) fprintf(file, "%d ", i);
    fclose(file);

    // check t_neigh_tri
    file = fopen("test_mesh3d.neightri", "w");
    for (int i=0;i<S.nt;i++) {
        for (int j=0;j<S.t_nb_neigh_tri[i];j++)
            fprintf(file, "%d ", S.t_neigh_tri[S.t_maxtri*i+j]);
        fprintf(file, "\n");
    }
    fclose(file);

    // check bounding sphere
    /*file = fopen("test_mesh3d.spheres", "w");
    for (int i=0;i<S.nt;i++) {
        double *c = S.bounding_sphere + 4*i;
        fprintf(file, "%.4f %.4f %.4f  %.4f\n", c[0], c[1], c[2], c[3]);
    }
    fclose(file);*/
}